### DISCUSSION TITLE
Overriding a Haskell package dependency version

### DISCUSSION THREAD

**Discussion by wolverian** [srid/haskell-flake#28] [Q&A] [answered]
I'm trying to use GHC 9.4.2 to compile my project, i.e. `haskellPackages = pkgs.haskell.packages.ghc942;`. However, at the moment, nixpkgs is giving me the version 0.5.1 of hspec-contrib, which doesn't compile on 9.4.2, so I'd like to override its version to 0.5.1.1, which should compile. I've tried:

```nix
        haskellProjects.default = {
          root = ./.;
          haskellPackages = pkgs.haskell.packages.ghc942;
          source-overrides = {
            hspec-contrib = "0.5.1.1";
          };
        };
```

However, this gives me the following error:

```
$ nix develop
[...]
error: A definition for option `perSystem.aarch64-darwin.haskellProjects.default.source-overrides.hspec-contrib' is not of type `path'. Definition values:
       - In `<unknown-file>, via option perSystem': "0.5.1.1"
(use '--show-trace' to show detailed location information)
```

Any clue how to fix this? Thanks!

**Comment by wolverian** [comment]
BTW, this is particularly confusing because my cabal file doesn't depend on hspec or hspec-contrib. The only dependency is `base`.

**Comment by srid** [comment, marked as answer]
Note to self: I should explain `source-overrides` in https://srid.ca/haskell-flake/dependency

Unlike `overrides`, when using `source-overrides` you are required to specify a Nix `path` value. This could be a flake input (with `flake = false;`); see "ema", "heist-extra" and "tailwind" in https://github.com/EmaApps/emanote/blob/master/flake.nix as an example.

That said, *if* you are trying to use a specific version on Hackage, you should instead use `overrides` and [`callHackage`](https://tek.brick.do/how-to-override-dependency-versions-when-building-a-haskell-project-with-nix-K3VXJd8mEKO7):

```nix
haskellProjects.default = 
  overrides = self: super: {
    hspec-contrib = self.callHackage "hspec-contrib" "0.5.1.1" {};
  };
};
```

By the way, you can also use this overlay directly if you want GHC 9.4 support: https://horizon-haskell.net/ ... see https://github.com/srid/haskell-template/pull/55

**Reply by wolverian** [reply]
Thank you, this clarifies things. I was confused by the documentation of `source-overrides` in Nixpkgs Haskell code I think.

**Reply by wolverian** [reply]
Unfortunately Horizon doesn't provide aarch64-darwin packages, which are what I'm after.

**Reply by srid** [reply]
cc @locallycompact
